src/smpp_socket.erl

```diff
@@ -955,3 +976,52 @@ pp(submit_sm_resp, _) -> record_info(fields, submit_sm_resp);
 pp(unbind, _) -> record_info(fields, unbind);
 pp(unbind_resp, _) -> record_info(fields, unbind_resp);
 pp(_, _) -> no.
+
+%%%-------------------------------------------------------------------
+%%% Overload protection
+%%%-------------------------------------------------------------------
+-spec init_rate_limit(state()) -> state().
+init_rate_limit(#{req_per_sec := undefined} = State) ->
+    State;
+init_rate_limit(#{req_per_sec := Rps} = State) ->
+    % keep one slot for eventual enquire_link request
+    State#{req_per_sec := Rps - 1, requests => []}.
+
+-spec check_limits(state()) ->
+          {ok, state()} |
+          {error, {statename(), state(), [gen_statem:action()]}}.
+check_limits(#{req_per_sec := undefined} = State) ->
+    check_in_flight(State);
+check_limits(#{req_per_sec := Rps} = State0) ->
+    Now = erlang:monotonic_time(millisecond),
+    case add_request(State0, Now div 1000) of
```

Comment:
Supervisor intensity logic is an overkill in this case. Supervisor intensity is measured during a period of _one to infinity seconds_, whilst SMPP rate limit is measured during _exactly one second_. I.e. this array manipulation is way too ambiguous for the task here.

It's enough to have something like this:
```erlang
check_rate_limit(#{req_per_sec := undefined} = State) ->
    check_in_flight_limit(State);
check_rate_limit(#{req_per_sec := MaxRate, current_rps := {ActiveSecond, Rate}} = State0) ->
    case erlang:monotonic_time(second) of
        ActiveSecond when Rate == MaxRate ->
            NowMs = erlang:monotonic_time(millisecond),
            Timeout = max(0, round((ActiveSecond + 1) * 1000 - NowMs)),
            {error, {rate_limit_cooldown,
                         State0,
                         [postpone, {state_timeout, Timeout, rate_limit_reset}]}};
        ActiveSecond ->
            check_in_flight_limit(State0#{current_rps := {ActiveSecond, Rate + 1}});
        Now ->
            check_in_flight_limit(State0#{current_rps := {Now, 1}})
    end.
```

Reply:
fixed